sw/device/lib/crypto/drivers/rv_core_ibex.c

```diff
@@ -0,0 +1,37 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/drivers/rv_core_ibex.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_core_ibex_regs.h"
+
+enum {
+  kBaseAddr = TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR,
+};
+
+/**
+ * Blocks until data is ready in the RND register.
+ */
+static void wait_rnd_valid(void) {
+  while (true) {
+    uint32_t reg =
+        abs_mmio_read32(kBaseAddr + RV_CORE_IBEX_RND_STATUS_REG_OFFSET);
+    if (bitfield_bit32_read(reg, RV_CORE_IBEX_RND_STATUS_RND_DATA_VALID_BIT)) {
+      return;
+    }
+  }
+}
+
+uint32_t ibex_rnd32_read(void) {
+  wait_rnd_valid();
+  return abs_mmio_read32(kBaseAddr + RV_CORE_IBEX_RND_DATA_REG_OFFSET);
```

Comment:
nit: low priority comment. Do you think we should also loop if the read value is 0? This is to account for the unlikely event of an interrupt reading this register, which would cause this read to return 0, or to account for faults. 

For the fault case we'd have to check the assembly to see if the loop would be helpful. I suppose we could add a hardened check at the call site for cases where it would be important to check.

Reply:
Hmm, interesting. Retrying if we read zero might also be a good way to add a "timeout" functionality -- it makes me a little uncomfortable that this just loops until entropy is available, but we could then add a maximum loop iteration count and return zero in that case. Then we'd tell callers they need to handle zero as an error, which would let programs fail more gracefully if entropy isn't available.

For now I'll not block merging on this but it's a good idea for an improvement.

Reply:
Sounds good. I think we'll have to revisit the return 0 later on.